### PR TITLE
lammps: update, hoomd-blue, dl_poly: init

### PR DESCRIPTION
**Pull request by costrouc**
###### Motivation for this change

Adding molecular dynamics codes that our group uses at UTK. Hoomd-blue is the most interesting addition. While I have not added GPU support it is mpi enabled and a fun code to work with. I do not have enough experience packaging with GPUs yet (maybe will be packaged at a future date). Scales to 10K GPUs according to benchmarks.

Would like to request @markuskowa for review.

###### Things done

lammps: patch_2Aug2018 -> stable_22Aug2018
dl-poly-classic: init 1.10
hoomd-blue: init at 2.3.4 

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [X] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/applications/science/molecular-dynamics/hoomd-blue/default.nix [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+  src = fetchgit {
+    url = "https://bitbucket.org/glotzer/hoomd-blue";
+    rev = "v${version}";
+    sha256 = "0in49f1dvah33nl5n2qqbssfynb31pw1ds07j8ziryk9w252j1al";
+  };
+
+  passthru = {
+    inherit components mpi;
+  };
```
Comment (markuskowa) [inline pkgs/applications/science/molecular-dynamics/hoomd-blue/default.nix]: What is the purpose of `components`?

File: pkgs/applications/science/molecular-dynamics/hoomd-blue/default.nix [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+  buildInputs = [ cmake stdenv pkgconfig ]
```
Comment (markuskowa) [inline pkgs/applications/science/molecular-dynamics/hoomd-blue/default.nix]: `pkgconfig` and `cmake` should be in `nativeBuildInputs`. No need to put `stdenv` to build inputs.
 The line should then read `nativeBuildInputs = [ cmake pkgconfig ];`
mpi goes into `buildInputs`.

File: pkgs/applications/science/molecular-dynamics/hoomd-blue/default.nix [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+       "-DBUILD_CGCMM="${onOffBool components.cgcmm}"
+       "-DBUILD_DEPRECIATED="${onOffBool components.depreciated}"
+       "-DBUILD_HPMC="${onOffBool components.hpmc}"
+       "-DBUILD_MD="${onOffBool components.md}"
+       "-DBUILD_METAL="${onOffBool components.metal}"
+       "-DBUILD_TESTING="${onOffBool components.testing}"
+    )
+  '';
+
+  checkTarget = "test";
```
Comment (markuskowa) [inline pkgs/applications/science/molecular-dynamics/hoomd-blue/default.nix]: That only has an effect if you set `doCheck=true;`

File: pkgs/applications/science/molecular-dynamics/lammps/default.nix [hunk 4]
```diff
@@ -67,8 +61,8 @@ stdenv.mkDerivation rec {
       under the terms of the GNU Public License (GPL).
       '';
     homepage = http://lammps.sandia.gov;
-    license = stdenv.lib.licenses.gpl2;
-    platforms = stdenv.lib.platforms.linux;
-    maintainers = with lib.maintainers; [ costrouc ];
+    license = licenses.gpl2;
+    platforms = [ "x86_64-linux" ];
```
Comment (markuskowa) [inline pkgs/applications/science/molecular-dynamics/lammps/default.nix]: The previous version actual did build on `aarch64`. I guess we could leave it as `platforms.linux` and see if it builds on arm.

File: pkgs/applications/science/molecular-dynamics/lammps/default.nix [hunk 5]
```diff
@@ -1,63 +1,57 @@
...
-    make mode=shlib ${if mpiSupport then "mpi" else "serial"} SHELL=$SHELL LMP_INC="${lammps_includes}" FFT_PATH=-DFFT_FFTW3 FFT_LIB=-lfftw3 JPG_LIB=-lpng
+  # Must do manual build due to LAMMPS requiring a seperate build for
+  # the libraries and executable. Also non-typical make script
+  buildPhase = ''
+    make mode=exe ${if (mpi != null) then "mpi" else "serial"} SHELL=$SHELL LMP_INC="${lammps_includes}" FFT_PATH=-DFFT_FFTW3 FFT_LIB=-lfftw3 JPG_LIB=-lpng
+    make mode=shlib ${if (mpi != null) then "mpi" else "serial"} SHELL=$SHELL LMP_INC="${lammps_includes}" FFT_PATH=-DFFT_FFTW3 FFT_LIB=-lfftw3 JPG_LIB=-lpng
+  '';
 
+  installPhase = ''
     mkdir -p $out/bin
```
Comment (markuskowa) [inline pkgs/applications/science/molecular-dynamics/lammps/default.nix]: A single `mkdir` line should also work here: `mkdir -p $out/bin $out/include $out/lib`.

**Comment by markuskowa** [comment]
As general comment (for the future): One could break that PR up into two or three PRs (at least the update as separate PR).

**Comment by costrouc** [comment]
@markuskowa the changes have been made. The reason why I am responding to the conversations on the code review in a single comment. Is that when I rebase with git (to keep it to only a few commits). Github completely looses track of where you made the comment. Is there a way around this?

> What is the purpose of `components`?

Components in hoomd-blue are a set of components/features to include with the installation. The components are included within the git repository. Here is the documenation for one of the componenets https://hoomd-blue.readthedocs.io/en/stable/package-hpmc.html.

**Comment by markuskowa** [comment]
@GrahamcOfBorg build lammps lammps-mpi dl-poly-classic-mpi pythonPackages36.hoomd-blue pythonPackages36.hoomd-blue-mpi

File: pkgs/top-level/all-packages.nix [hunk 6]
```diff
@@ -20894,14 +20894,19 @@ with pkgs;
 
   ### SCIENCE/MOLECULAR-DYNAMICS
 
+  dl-poly-classic-mpi = callPackage ../applications/science/molecular-dynamics/dl-poly-classic {
+    mpi = openmpi;
+  };
+
+  # hoomd-blue in python-packages.nix
```
Comment (markuskowa) [inline pkgs/top-level/all-packages.nix]: That line can be deleted?

**Comment by markuskowa** [comment]
@GrahamcOfBorg build python36Packages.hoomd-blue, python36Packages.hoomd-blue-mpi

File: pkgs/applications/science/molecular-dynamics/hoomd-blue/default.nix [hunk 7]
```diff
@@ -0,0 +1,62 @@
+{ stdenv, fetchgit
+, cmake, pkgconfig
+, python, numpy
```
Comment (FRidh) [inline pkgs/applications/science/molecular-dynamics/hoomd-blue/default.nix]: Don't pass in attributes of the Python package set. Instead pass in `pythonPackages` and use its attributes, or pass in `python`, and use `python.pkgs`' attributes.

File: pkgs/applications/science/molecular-dynamics/hoomd-blue/default.nix [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+  nativeBuildInputs = [ cmake pkgconfig ];
+  buildInputs = stdenv.lib.optionals (mpi != null) [ mpi ];
+  propagatedBuildInputs = [ python numpy ]
+   ++ stdenv.lib.optionals (mpi != null) [ mpi4py ];
+
+  enableParallelBuilding = true;
+
+  dontAddPrefix = true;
+  preConfigure = ''
+    cmakeFlagsArray=(
```
Comment (FRidh) [inline pkgs/applications/science/molecular-dynamics/hoomd-blue/default.nix]: There is a parameter, `cmakeFlags = [ ... ]`

File: pkgs/applications/science/molecular-dynamics/hoomd-blue/default.nix [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    rev = "v${version}";
+    sha256 = "0in49f1dvah33nl5n2qqbssfynb31pw1ds07j8ziryk9w252j1al";
+  };
+
+  passthru = {
+    inherit components mpi;
+  };
+
+  nativeBuildInputs = [ cmake pkgconfig ];
+  buildInputs = stdenv.lib.optionals (mpi != null) [ mpi ];
```
Comment (FRidh) [inline pkgs/applications/science/molecular-dynamics/hoomd-blue/default.nix]: Use a parameter, such as `withMpi`.

File: pkgs/top-level/all-packages.nix [hunk 10]
```diff
@@ -20894,14 +20894,19 @@ with pkgs;
...
+
   lammps = callPackage ../applications/science/molecular-dynamics/lammps {
     fftw = fftw;
   };
 
-  lammps-mpi = appendToName "mpi" (lammps.override {
-    mpiSupport = true;
+  lammps-mpi = callPackage ../applications/science/molecular-dynamics/lammps {
```
Comment (FRidh) [inline pkgs/top-level/all-packages.nix]: Don't call again but use override.

File: pkgs/top-level/python-packages.nix [hunk 11]
```diff
@@ -322,6 +322,12 @@ in {
 
   habanero = callPackage ../development/python-modules/habanero { };
 
+  hoomd-blue = callPackage ../applications/science/molecular-dynamics/hoomd-blue { };
+
+  hoomd-blue-mpi = callPackage ../applications/science/molecular-dynamics/hoomd-blue {
```
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: Do we need this attribute when one can also just override?

**Review by FRidh — CHANGES_REQUESTED** [review]

**Comment by costrouc** [comment]
All requested changes have been made. I have tested that hoomd works properly with a few of the example scripts. https://nbviewer.jupyter.org/github/joaander/hoomd-examples/blob/master/index.ipynb

**Comment by costrouc** [comment]
Rebased. @markuskowa with your approval this is ready for a merge. We will need to test rebuilding of `python36Packages.hoomd-blue` and `python36Packages.hoomd-blue-mpi`. I have verified that both build on my machine with sandbox. Hoomd tests were turned off but I can vouch that the program does run (with and without mpi) and was able to run several examples without errors.

**Comment by markuskowa** [comment]
I guess it needs another rebase.

**Comment by costrouc** [comment]
Wow a merge conflict in less than 8 hours, that's some good luck. Fixed now.

**Comment by costrouc** [comment]
@GrahamcOfBorg build python36Packages.hoomd-blue

**Comment by markuskowa** [comment]
@FRidh Are you OK with the changes?

**Comment by costrouc** [comment]
Bump. Ready for merge following your approval @FRidh. I have verified all of the packages to build properly.

File: pkgs/top-level/python-packages.nix [hunk 12]
```diff
@@ -366,6 +366,14 @@ in {
 
   hdmedians = callPackage ../development/python-modules/hdmedians { };
 
+  hoomd-blue = callPackage ../applications/science/molecular-dynamics/hoomd-blue {
```
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: Python packages need to live under `pkgs/development/python-modules`. If the package instead provide as well Python bindings, then the version in `all-packages.nix` needs to be overridden and `toPythonModule` should be used. Using `stdenv.mkDerivation` directly in `python-packages.nix` is not permitted.

File: pkgs/top-level/python-packages.nix [hunk 13]
```diff
@@ -366,6 +366,14 @@ in {
...
+    inherit python;
+  };
+
+  hoomd-blue-mpi = (self.hoomd-blue.override {
```
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: Let's drop this one. We don't include all possible ways something can be build because of limited CI capacity.

**Review by FRidh — CHANGES_REQUESTED** [review]

File: pkgs/top-level/python-packages.nix (hunk 13 above)
Comment (costrouc) [reply pkgs/top-level/python-packages.nix]: Fixed. I have droped hoomd-blue-mpi from attributes.

File: pkgs/top-level/python-packages.nix (hunk 12 above)
Comment (costrouc) [reply pkgs/top-level/python-packages.nix]: Fixed. I have used `toPythonModule` and the package has been moved to `python-modules`.

**Comment by costrouc** [comment]
@FRidh changes have been made and ready for merge.

**Comment by costrouc** [comment]
Bump. Ready for merge.

**Pull request merged**